### DISCUSSION TITLE
General AI understanding

### DISCUSSION THREAD

**Discussion by danlangford** [pappde/bmai#4] [General]
We can dedicate this thread to general AI questions. leave thread #2 for further "welcome" and "introductions" should they be needed. 

so digging into your explanation in #2 ...
> BMAI is an example of the type of AI you might use for a board game or traditional classic game. There is a discrete search space, and you want to pick the move that gives you the best expected value, whether by simulating out the entire game to the conclusion or by a heuristic function to judge the goodness of a final position in your search space. 

so if i were to try to dissect this for my understanding i read a couple approaches in here. let me know if i am understanding this correctly. when building an AI for a classic game you could:
1. search the entire space for a win and then make the moves that progresses you towards that win. this seems good for games in which there is 1 final win condition. Like a jigsaw puzzle or a sudoku.
2. search the entire space for all possible wins and then make moves that progress you towards the fastest or strongest or safest win depending on your design goals and the game. games like rock-paper-scissors have many possible outcomes but that entire space would be relatively easy to search. maybe black-jack is also a game with a simpler search space. 
3. search enough of the space to make the next move with high confidence. this is useful in games where the entire search space may not be feasible or possible to search. its helpful if in these games you have indicators of the goodness of the position you are left in. in a game like MtG you want to reduce their health to 0 before they reduce your to zero. So your search space could look for possible outcomes that leave them with fewer, and you with more, Health. 

and in regards to point 3 there "heuristics" describes how you judge the goodness of the final position your search space found? and i assume heuristics could also apply to point 2 to determine which of the multiple win paths had the highest "goodness"? (i like the word goodness as a unit of measurement) you mentioned opponent modeling maybe there is a way to observe that your opponent statistically favors "rock" in rock-paper-scissors which may help you determine that "paper" moves may increase the goodness of your decisions?

does BMAI model possible outcomes with each possible attack as well as each possible reroll? I assume that it does not search the entire tree then and uses point 3 to determine the final position of the chosen search space with the most goodness? is goodness judged by number of "sides" taken from the opponent? does it also consider the strength of the final position even if side count is low? Does it work on probabilities like the odds of somebody rerolling into a space that you can attack them? or does it simply evaluate each possible roll individually? does it model to a final positions in the search space and play towards the best final position or does it find with next move that goes down the tree with the most good final positions?

**Comment by pappde** [comment]
point 1: 
Looking at the entire search space may not be practical depending on the game. Also, I wouldn't use a puzzle as an example since - while that is a search problem - there are a small number of valid moves and all valid moves get you to the solution. Sudoku could be used as an example, but it is just a 1 player game with one decision maker. It's better to think about classic two (or more) player games such as tic-tac-toe, checkers, backgammon, mancala, ...

A typical approach is minimax search - at each alternating level of the search:
- Your Turn: you want to choose the move that **max**imizes your chance of winning 
- Opponent Turn: your opponent(s) will then choose the move that maximizes their chance of winning (generally the same as **mini**mizing your chance of winning)
- Your Turn: repeat

This analysis will give you the highest value move in the current position. In fact, a lot of techniques that go into minimax algorithms is about "pruning" the search tree (quickly eliminating moves that are obviously bad) so you can go deeper. 

Most likely, what you'll be able to get is to a to several "future states in the game" with a computed/estimated "probably of win from this position".  You want to pick the move that gives you the maximum "probability of winning" from the current position.

You'll note that this approach requires a way to score moves from the opponent's point of view. Normally, you would just use the same model as your decision making, but this is an opportunity to work in some opponent modeling.

point 2: 
Rock-paper-scissors is a two move game only (your move + your opponent's move). There's really no search - it's all about opponent modeling. Your example is a good one - if you observe your opponent favors rock, then favoring paper is an maximal strategy.

Also, blackjack is a really a one player game, so I wouldn't use that as an example either. You definitely want to think two player games, which makes it a little difficult to think about it as "find all possible wins".

point 3:
MtG is a good example where you can compute a "goodness" for a position without simulating out the entire game.  I believe BMAI may leverage this in some cases.  However, another way to compute a probably of winning for a position is to simulate out the entire game (monte carlo simulations). This does require having a quick decision mechanism for future positions.  

When you are doing your search in a game with randomness, here are some ways you can score a position:
1) Heuristic (fastest): using some O(1) function score the game state like my total dice vs your total dice.
2) Simulations with simple decision making: simulate playing the game out 100 times with a quick decision making system for deeper nodes.
3) Expanding the search using something like minimax to pick the best move. BMAI code uses the term "ply" to refer to the next level - then in the next level. Also, to make things more complex you still need to factor in the die rolls.

the third method will give you the most accurate answer, but is the most expensive since it is exponential.  In fact, I believe BMAI computes the branching factor at each node (total number of moves), and may choose which method to use depending on that.

If you visualize the search as a tree, you have nodes (game states) and branches (move taken + dice roll result).  If a node is a "victory" then you have your score - 0 or 1.   If you use the heuristic at a node, then you have some quickly computed value (e.g. 0.75).  If you use the simulations, then your node score is going to be an average of the score of the child nodes. In the third case, your score is going to be the minimax of the child nodes.

BMAI factors in rerolls. Simple example: D4 D4 vs D4 D4, 
- Initial State: "3 1" vs "3 2"
- Level 0: you have 2 valid moves:
Move 1 = 3 captures 3, 
Move 2 = 3 captures 2.
In each case, there are 4 different outputs for the reroll.
--- Level 1: there are 8 possible nodes
Move 1 Child Nodes: "4 1" vs "2", "3 1" vs "2", "2 1" vs "2", "1 1 vs 2", each with a 25% chance
Move 2 Child Nodes: "4 1" vs "3", "3 1" vs "3", "2 1" vs "3", "1 1 vs 3", each with a 25% chance

At this point, for each of these nodes (game states), you can choose how to compute a score, by either heuristic, simple simulations, or expanding the minimax search.  Let's use the heuristic method. Here, we compute "Probably Win (P) = your total dice / all total dice"
Move 1 Child Nodes: 5/7, 4/6, 3/5, 2/4
Move 2 Child Nodes: 5/8, 4/7, 3/6, 2/5

Since each child state has a 25% chance of occurring, we can therefore compute the P for the Level 0 Moves:
P(Move1) = 62%
P(Move2) = 41%

Therefore Move1 is best.  Note that this is not illustrating the idea of simulating out the game to compute a more accurate "P" or doing the minimax expansion, but hopefully is enough detail to be able to extrapolate.

You can see that ways in which to make the AI better:
1) better heuristic function.  You could even "cache" game states to a database (which is a technique that has been used in large search space games like checkers)
2) smarter quick decision AI for simulating
3) going deeper in the minimax search (since this makes the best decisions)

**Comment by danlangford** [comment]
this is every helpful thank you so much. what i am seeing in the logs is lining up with these concepts here and helping it all click together a bit for me. Visualizing it also helps me so i tried to build out the tree in your example.

![image](https://user-images.githubusercontent.com/721364/103010928-54c87900-44f6-11eb-9ced-246278692ef1.png)

 i can see how important a good heuristic function is. and as a player of the game i immediately said "well the ratio of the total sides that i control is not always a good indicator of the strength of my position... it would be better to determine how many moves the over player has against me and the strength of those moves... ohhhhhhh" and right then the minimax approach really made sense.

**Comment by pappde** [comment]
You can also immediately see how this particular heuristic is weak - for example, it doesn't factor in the dice sides only the face, doesn't factor in different types of dice, doesn't factor in current score.  In fact, this is a wonderful example of great application of machine learning - input the game state, output the score. Over time with enough training you could have a very powerful move evaluator. I would love to do a project like that.

The example is a great illustration of combining the concepts of action selection with random state, but it could be contrived a little better to demonstrate minimax - Move 1 puts you in positions that score higher for the heuristic but leaves you exposed for your opponent to then do a move that gives them a lot of points. E.g. if you have 20:6 6:6 and use your 20 to capture. When searching "one ply deeper" you are repeating the above analysis, but from the point of view of the opponent. So you expand a node out into another tree.

In the end, the AI has a few different options for scoring/evaluating the goodness of any given node.
